src/mo_frontend/typing.ml

```diff
@@ -1150,7 +1150,18 @@ and infer_exp'' env exp : T.typ =
         { env with async = C.NullCap; in_actor = true }
       else env
     in
-    infer_obj env' obj_sort.it dec_fields exp.at
+    let t = infer_obj env' obj_sort.it dec_fields exp.at in
+    begin match typ_opt with
+      | Some typ ->
+        let t' = check_typ env' typ in
+        if not (T.sub t t') then
+          local_error env exp.at "M0192"
+            "body of type%a\ndoes not match expected type%a"
```

Comment:
I don't try to spell out which sort of body it is, as it should be obvious from the inferred type:
``` Motoko
stdin: type error [M0192], body of type
  module {go : () -> ()}
does not match expected type
  module {got : () -> ()}
```